src/opensles/AudioStreamOpenSLES.cpp

```diff
@@ -129,10 +133,10 @@ SLresult AudioStreamOpenSLES::enqueueCallbackBuffer(SLAndroidSimpleBufferQueueIt
 
 SLresult AudioStreamOpenSLES::processBufferCallback(SLAndroidSimpleBufferQueueItf bq) {
     // Ask the callback to fill the output buffer with data.
-    Result result = fireCallback(mCallbackBuffer, mFramesPerCallback);
-    if (result != Result::OK) {
+    DataCallbackResult result = fireCallback(mCallbackBuffer, mFramesPerCallback);
+    if (result != DataCallbackResult::Continue) {
         LOGE("Oboe callback returned %d", result);
-        return SL_RESULT_INTERNAL_ERROR;
+        return SL_RESULT_INTERNAL_ERROR; // TODO How should we stop OpenSL ES.
```

Comment:
We could use the SL player interface to set the play state to stopped. E.g. 

https://github.com/googlesamples/android-ndk/blob/master/audio-echo/app/src/main/cpp/audio_player.cpp#L235

Additionally, we could set the stream state to stopped.

Reply:
Can that be called safely from inside an OpenSL callback?

Reply:
90% sure they can. We should probably add a test for this.

Reply:
This is a separate issue from the read/write patch. 
I filed:   https://github.com/google/oboe/issues/61

Let's finish up this patch.